src/internet/spotifywebapi/spotifywebapiservice.cpp

```diff
@@ -0,0 +1,197 @@
+/* This file is part of Clementine.
+   Copyright 2021, Kenman Tsang <kentsangkm@pm.me>
+
+   Clementine is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   Clementine is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with Clementine.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include "spotifywebapiservice.h"
+
+#include <QJsonArray>
+#include <QJsonDocument>
+#include <QJsonObject>
+#include <QtDebug>
+#include <utility>
+
+#include "3rdparty/qtiocompressor/qtiocompressor.h"
+#include "core/application.h"
+#include "core/network.h"
+#include "core/timeconstants.h"
+#include "globalsearch/globalsearch.h"
+#include "globalsearch/spotifywebapisearchprovider.h"
+#include "ui/iconloader.h"
+
+namespace {
+static constexpr const char* kServiceName = "SpotifyWebApi";
+static constexpr const char* kGetAccessTokenUrl =
+    "https://open.spotify.com/"
+    "get_access_token?reason=transport&productType=web_player";
+static constexpr const char* kSearchUrl =
+    "https://api.spotify.com/v1/search?q=%1&type=track&limit=50";
+
+template <typename... Args>
+inline QJsonValue Get(QJsonValue obj, Args&&... args) {
+  std::array<const char*, sizeof...(Args)> names = {
+      std::forward<Args>(args)...};
+  for (auto&& name : names) {
+    Q_ASSERT(obj.isObject());
+    obj = obj.toObject()[name];
+  }
+  return obj;
+}
+
+template <typename... Args>
+inline QJsonValue Get(const QJsonDocument& obj, Args&&... args) {
+  return Get(obj.object(), std::forward<Args>(args)...);
+}
+
+QString concat(const QJsonArray& array, const char* name) {
+  QStringList ret;
+  for (auto&& item : array) {
+    ret << Get(item, name).toString();
+  }
+  return ret.join(", ");
+}
+
+}  // namespace
+
+SpotifyWebApiService::SpotifyWebApiService(Application* app,
+                                           InternetModel* parent)
+    : InternetService(kServiceName, app, parent, parent),
+      network_(new NetworkAccessManager{this}) {
+  app_->global_search()->AddProvider(
+      new SpotifyWebApiSearchProvider(app_, this));
+}
+
+SpotifyWebApiService::~SpotifyWebApiService() {}
+
+QStandardItem* SpotifyWebApiService::CreateRootItem() {
+  root_ = new QStandardItem(IconLoader::Load("spotify", IconLoader::Provider),
+                            kServiceName);
+  return root_;
+}
+
+void SpotifyWebApiService::LazyPopulate(QStandardItem* item) {}
+
+void SpotifyWebApiService::Search(int searchId, QString queryStr) {
+  if (std::chrono::system_clock::now().time_since_epoch() >=
```

Comment:
Use: QDateTime::currentDateTime().toSecsSinceEpoch()

Reply:
thx for the tips. Fixed